source/slang/slang-preprocessor.cpp

```diff
@@ -896,280 +1287,763 @@ static Index _parseMacroArgs(
+                {
+                    // Note that `tokenOpIndex` is being set here to the op index for the
+                    // right-hand operand to the `##`. This is appropriate for cases where
+                    // you might have chained `##` ops:
+                    //
+                    //      #define F(X,Y,Z) X ## Y ## Z
+                    //
+                    // If `Y` expands to a single token, then `X ## Y` should be treated
+                    // as the left operand to the `Y ## Z` paste.
+                    //
+                    token = m_currentOpStreams.readToken();
+                    tokenOpIndex = m_macroOpIndex;
+                }
 
-                // Add to the start of the stream
-                SimpleTokenInputStream* simpleStream = createSimpleInputStream(preprocessor, newToken);
-                PushInputStream(preprocessor, simpleStream);
-                break;
+                // At this point we are ready to head back to the top of the loop and see
+                // if our invariants have been re-established.
             }
+            break;
         }
     }
 }
 
-// Read one token with macro-expansion enabled.
-static Token AdvanceToken(Preprocessor* preprocessor)
+void MacroInvocation::_pushSingleTokenStream(TokenType tokenType, SourceLoc tokenLoc, UnownedStringSlice const& content)
 {
-top:
-    // Check whether we need to macro expand at the cursor.
-    MaybeBeginMacroExpansion(preprocessor);
+    // The goal here is to push a token stream that represents a single token
+    // with exactly the given `content`, etc.
+    //
+    // We are going to keep the content alive using the slice pool for the source
+    // manager, which will also lead to it being shared if used multiple times.
+    //
+    SourceManager* sourceManager = m_preprocessor->getSourceManager();
+    auto& pool = sourceManager->getStringSlicePool();
+    auto poolHandle = pool.add(content);
+    auto slice = pool.getSlice(poolHandle);
 
-    // Read a raw token (now that expansion has been triggered)
-    Token token = AdvanceRawToken(preprocessor);
+    Token token;
+    token.type = tokenType;
+    token.setContent(slice);
+    token.loc = tokenLoc;
+
+    TokenList lexedTokens;
+    lexedTokens.add(token);
+
+    // Every token list needs to be terminated with an EOF,
+    // so we will construct one that matches the location
+    // for the `token`.
+    //
+    Token eofToken;
+    eofToken.type = TokenType::EndOfFile;
+    eofToken.loc = token.loc;
+    eofToken.flags = TokenFlag::AfterWhitespace | TokenFlag::AtStartOfLine;
+    lexedTokens.add(eofToken);
+
+    SingleUseInputStream* inputStream = new SingleUseInputStream(m_preprocessor, lexedTokens);
+    m_currentOpStreams.push(inputStream);
+}
 
-    // Check if we need to perform token pasting
-    if (PeekRawTokenType(preprocessor) != TokenType::PoundPound)
+template<typename F>
```

Comment:
Obviously this is down to a lot of stuff related to personal experiences with programming, so the following is not meant to be any kind of attempt to establish that I'm "right." I just want to outline my reasoning in case it is interesting...

First, just like you might have an allergic respond to a template, I tend to have an allergic response to the idiom of:

```
switch(thing) {
...
case A:
case B:
     /* shared code */
    if(thing == A) {
        /* A-specific code */
    } else {
        /* B-specific code */
    }
    /* more shared code */
}
```

There end up being two different conditional branch points on `thing`, and an invariant between them that needs to be respected. Whenever possible, I like to have things look more like:

```
switch(thing) {
...
case A: sharedCode(/* shared args */, /* A-specific args */ ); break;
case B: sharedCode(/* shared args */, /* B-specific args */ ); break;
...
}
```

This alternative keeps to a single dispatch on `thing`, while still avoiding the code duplication. Of course it isn't always as easy if the `A-specific args` needs to include more code/logic, which is how this case ended up in template-land.

Second, part of why I prefer the latter option above is because my experience is that "there's always another case." In both compilers and graphics code I tend to find that if I write an `if`, it eventually becomes a `switch` because you shift from two cases to three (that's part of why I often write `switch` statements where others might use an `if` - I am branch-predicting that the `switch` will be warranted later, and I'd like to avoid an `if` cascade if possible).

Similarly, if I see:

```
case A:
case B:
    ...
```

I pretty much assume that the list of cases that need similar handling is definitely going to grow at some point to include a third, so that things like an embedded `if(thing == A)` come across as a red flag to me (because that nested `if` will become a nested `switch(thing)` and that seems to be even more odd, right?).

In this *specific* case, I know that some other compilers (e.g., the Swift compiler) have three or more pre-defined things akin to `__FILE__` (although their syntax is different for that stuff). In practice, a user might want any of:

* The full (or full-est possible) path to the current file
* The path "stem" by which the file was found or identified (usually relative to the working directory of the compile)
* Just the file name, without directory information
* The shortest path/name that uniquely identifies a given file in the context of a particular compile

There are benefits to multiple options on this list, and it may not be that there is a good "one size fits all" option. As such, one could easily imagine Slang supporting both `__FILE__` and `__FILE_NAME__`, or even something like `__FILE_NAME_HASH__` that could be used in release builds of a project that want to include sufficient information to identify source locations in logs, but that don't reveal anything about actual filesystem paths on the machines that built a binary.